src/analyses/base.ml

```diff
@@ -2059,6 +2059,53 @@ struct
     let st: store = ctx.local in
     let gs = ctx.global in
     let desc = LF.find f in
+    let memory_copying dst src =
+      let dest_a, dest_typ = addr_type_of_exp dst in
+      let src_lval = mkMem ~addr:(Cil.stripCasts src) ~off:NoOffset in
+      let src_typ = eval_lv (Analyses.ask_of_ctx ctx) gs st src_lval
+                    |> AD.get_type in
+      (* when src and destination type coincide, take value from the source, otherwise use top *)
+      let value = if typeSig dest_typ = typeSig src_typ then
+          let src_cast_lval = mkMem ~addr:(Cilfacade.mkCast ~e:src ~newt:(TPtr (dest_typ, []))) ~off:NoOffset in
+          eval_rv (Analyses.ask_of_ctx ctx) gs st (Lval src_cast_lval)
+        else
+          VD.top_value (unrollType dest_typ)
+      in
+      set ~ctx (Analyses.ask_of_ctx ctx) gs st dest_a dest_typ value in
+    (* for string functions *)
+    let eval_n = function
+      (* if only n characters of a given string are needed, evaluate expression n to an integer option *) 
+      | Some n ->
+        begin match eval_rv (Analyses.ask_of_ctx ctx) gs st n with
+          | `Int i ->
+            begin match ID.to_int i with
+              | Some x -> Some (Z.to_int x)
+              | _ -> Some (-1)
+            end
+          | _ -> Some (-1)
+        end
+      (* do nothing if all characters are needed *)
+      | _ -> None 
```

Comment:
Why those `-1`s? Shouldn't they also be `None` if the integer is ambiguous or something?

Reply:
If they'd be `None`, then when the next function is called, e.g. `string_comparison`, then the equivalent to `strcmp` (without the `n`) would be performed as then instead of using `Addr.to_n_string`, `Addr.to_string` would be picked. However, when returning `Some (-1)`, then it's still `Addr.to_n_string` and the latter transforms negative values to `None`, which leads `string_comparions` to just return top